packetbeat/protos/dns/dns.go

```diff
@@ -707,3 +714,160 @@ func decodeDnsPacket(data []byte) (dns *layers.DNS, err error) {
 	}
 	return d, nil
 }
+
+// TCP implementation
+
+func (dns *Dns) Parse(pkt *protos.Packet, tcpTuple *common.TcpTuple, dir uint8, private protos.ProtocolData) protos.ProtocolData {
+	defer logp.Recover("DNS ParseTcp")
+
+	logp.Debug("dns", "Parsing packet addressed with %s of length %d.",
+		pkt.Tuple.String(), len(pkt.Payload))
+
+	priv := dnsPrivateData{}
+
+	if private != nil {
+		var ok bool
+		priv, ok = private.(dnsPrivateData)
+		if !ok {
+			priv = dnsPrivateData{}
+		}
+	}
+
+	payload := make([]byte, 0)
+
+	// Offset is critical
+	if len(pkt.Payload) > 2 {
+		payload = pkt.Payload[2:]
+	}
+
+	if priv.Data[dir] == nil {
+		priv.Data[dir] = &DnsStream{
+			tcpTuple: tcpTuple,
+			data:     payload,
+			message:  &DnsMessage{Ts: pkt.Ts, Tuple: pkt.Tuple},
+		}
+
+	} else {
+		priv.Data[dir].data = append(priv.Data[dir].data, payload...)
+		if len(priv.Data[dir].data) > tcp.TCP_MAX_DATA_IN_STREAM {
+			logp.Debug("dns", "Stream data too large, dropping DNS stream")
+			priv.Data[dir] = nil
+			return priv
+		}
+	}
+
+	stream := priv.Data[dir]
+	if stream.message == nil {
+		stream.message = &DnsMessage{Ts: pkt.Ts, Tuple: pkt.Tuple}
+	}
+
+	// what kind of checks should be done here ?
+	// gopacket decodeDns already check QDCount, ANCount, NScount, ARcount
+	data := dns.messageParser(stream)
+
+	if data == nil {
+		logp.Debug("dns", NonDnsCompleteMsg+" addresses %s, length %d",
+			tcpTuple.String(), len(stream.data))
+
+		// drop this tcp stream. Will retry parsing with the next
+		// segment in it
+		priv.Data[dir] = nil
+		return priv
+	}
+
+	dns.messageComplete(tcpTuple, dir, stream, data)
+
+	return priv
+}
+
+// return decoded data so we don't have to do decode twice
+// return nil if failed
+func (dns *Dns) messageParser(s *DnsStream) *layers.DNS {
```

Comment:
Consider removing this method. I think anywhere it is used you could easily call `decodeDnsData`. In any case it should be a function and not a method on the `*Dns` receiver since it doesn't appear to use `dns`.